File: web/src/services/api/attendance.ts
import { request } from './client';

export type AttendanceRecord = {
  id: string;
  date: string;
  status: 'present' | 'late' | 'absent';
  source: string;
  scan_uid: string | null;
  scanned_at: string | null;
  notes: string | null;
};

export type AttendanceOverview = {
  student_id: string;
  student_name: string;
  student_code: string;
  roll_no: string;
  grade: string;
  section: string;
  attendance_percentage: number;
  present_days: number;
  late_days: number;
  recorded_days: number;
  today_status: 'present' | 'late' | 'absent' | 'not_marked';
  latest_scan_at: string | null;
  records: AttendanceRecord[];
};

export type AttendanceCard = {
  uid: string;
  label: string | null;
  student_name: string;
  student_code: string;
  grade: string;
  section: string;
  active: boolean;
};

const today = new Date();
const daysAgo = (days: number) => {
  const value = new Date(today);
  value.setDate(today.getDate() - days);
  return value.toISOString();
};
const dateAgo = (days: number) => daysAgo(days).slice(0, 10);

export const demoAttendanceCards: AttendanceCard[] = [
  { uid: 'DEMO-NFC-001', label: 'Demo NFC Card - Demo Student', student_name: 'Demo Student', student_code: 'STU-DEMO-001', grade: 'Grade 12', section: 'A', active: true },
  { uid: 'DEMO-NFC-002', label: 'Demo NFC Card - Aarav Sharma', student_name: 'Aarav Sharma', student_code: 'STU-DEMO-002', grade: 'Grade 12', section: 'A', active: true },
  { uid: 'DEMO-NFC-003', label: 'Demo NFC Card - Isha Patel', student_name: 'Isha Patel', student_code: 'STU-DEMO-003', grade: 'Grade 12', section: 'A', active: true },
  { uid: 'DEMO-NFC-004', label: 'Demo NFC Card - Kabir Khan', student_name: 'Kabir Khan', student_code: 'STU-DEMO-004', grade: 'Grade 11', section: 'A', active: true },
  { uid: 'DEMO-NFC-005', label: 'Demo NFC Card - Meera Nair', student_name: 'Meera Nair', student_code: 'STU-DEMO-005', grade: 'Grade 10', section: 'B', active: true },
];

export const demoAttendanceRecords: AttendanceOverview[] = demoAttendanceCards.map((card, index) => {
  const records: AttendanceRecord[] = [
    { id: `${card.uid}-6`, date: dateAgo(6), status: 'present', source: 'demo', scan_uid: card.uid, scanned_at: daysAgo(6), notes: 'Demo attendance history' },
    { id: `${card.uid}-5`, date: dateAgo(5), status: 'late', source: 'demo', scan_uid: card.uid, scanned_at: daysAgo(5), notes: 'Demo attendance history' },
    { id: `${card.uid}-4`, date: dateAgo(4), status: 'present', source: 'demo', scan_uid: card.uid, scanned_at: daysAgo(4), notes: 'Demo attendance history' },
    { id: `${card.uid}-3`, date: dateAgo(3), status: 'absent', source: 'demo', scan_uid: null, scanned_at: daysAgo(3), notes: 'Demo attendance history' },
    { id: `${card.uid}-2`, date: dateAgo(2), status: 'present', source: 'demo', scan_uid: card.uid, scanned_at: daysAgo(2), notes: 'Demo attendance history' },
    { id: `${card.uid}-1`, date: dateAgo(1), status: 'present', source: 'demo', scan_uid: card.uid, scanned_at: daysAgo(1), notes: 'Demo attendance history' },
  ];
  const presentDays = records.filter(record => record.status === 'present').length;
  const lateDays = records.filter(record => record.status === 'late').length;
  return { student_id: `demo-${index + 1}`, student_name: card.student_name, student_code: card.student_code, roll_no: `${index + 1}`, grade: card.grade, section: card.section, attendance_percentage: Math.round(((presentDays + lateDays) / records.length) * 100), present_days: presentDays, late_days: lateDays, recorded_days: records.length, today_status: 'not_marked', latest_scan_at: records[0].scanned_at, records };
});

export function applyDemoScan(cardUid: string) {
  const card = demoAttendanceCards.find(item => item.uid === cardUid);
  const student = demoAttendanceRecords.find(item => item.student_code === card?.student_code);
  if (!card || !student) throw new Error('Demo NFC card not found');
  const scannedAt = new Date().toISOString();
  const todayDate = scannedAt.slice(0, 10);
  const existing = student.records.find(record => record.date === todayDate);
  if (existing) {
    existing.status = 'present';
    existing.scanned_at = scannedAt;
    existing.scan_uid = cardUid;
    existing.notes = 'Demo scan updated from UI';
  } else {
    student.records.unshift({ id: `${cardUid}-today`, date: todayDate, status: 'present', source: 'demo', scan_uid: cardUid, scanned_at: scannedAt, notes: 'Demo scan recorded from UI' });
  }
  student.today_status = 'present';
  student.latest_scan_at = scannedAt;
  student.present_days = student.records.filter(record => record.status === 'present').length;
  student.late_days = student.records.filter(record => record.status === 'late').length;
  student.recorded_days = student.records.length;
  student.attendance_percentage = Math.round(((student.present_days + student.late_days) / student.recorded_days) * 100);
  return student;
}

export async function getAttendanceOverview(studentCode = 'STU-DEMO-001') {
  return request<AttendanceOverview>(`/attendance/overview?student_code=${encodeURIComponent(studentCode)}`);
}

export async function getAttendanceRecords() {
  return request<AttendanceOverview[]>('/attendance/records');
}

export async function getAttendanceCards() {
  return request<AttendanceCard[]>('/attendance/cards');
}

export async function scanAttendance(cardUid: string, deviceLabel = 'Mobile NFC') {
  return request<{ recorded: boolean; created: boolean; message: string; attendance: AttendanceOverview }>('/attendance/scan', { method: 'POST', body: JSON.stringify({ card_uid: cardUid, device_label: deviceLabel }) });
}

File: backend/app/api/v1/routes/attendance.py
from datetime import date, datetime, timezone
from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.models import AttendanceRecord, AttendanceStatus, ClassRoom, RFIDCard, StudentProfile, User, UserRole

router = APIRouter()


class NFCScanPayload(BaseModel):
    card_uid: str = Field(min_length=2, max_length=120)
    device_label: str | None = Field(default=None, max_length=120)


def iso(value: datetime | None) -> str | None:
    return value.isoformat() if value else None


async def student_rows(db: AsyncSession, student_code: str | None = None, grade: str | None = None, section: str | None = None) -> list[tuple[StudentProfile, User, ClassRoom]]:
    query = select(StudentProfile, User, ClassRoom).join(User, User.id == StudentProfile.user_id).join(ClassRoom, ClassRoom.id == StudentProfile.class_id).where(User.role == UserRole.student)
    if student_code:
        query = query.where(StudentProfile.student_code == student_code)
    if grade:
        grade_options = [grade, f"Grade {grade}"] if not grade.lower().startswith("grade") else [grade, grade.replace("Grade ", "")]
        query = query.where(ClassRoom.grade.in_(grade_options))
    if section:
        query = query.where(ClassRoom.section == section)
    return list((await db.execute(query.order_by(ClassRoom.grade.desc(), ClassRoom.section, StudentProfile.roll_no))).all())


async def serialize_student_attendance(db: AsyncSession, profile: StudentProfile, account: User, classroom: ClassRoom) -> dict:
    records = list((await db.scalars(select(AttendanceRecord).where(AttendanceRecord.student_id == profile.user_id).order_by(AttendanceRecord.attendance_date.desc(), AttendanceRecord.scanned_at.desc()))).all())
    present = sum(1 for record in records if record.status == AttendanceStatus.present)
    late = sum(1 for record in records if record.status == AttendanceStatus.late)
    total = len(records)
    percentage = round(((present + late) / total) * 100, 2) if total else float(profile.attendance_pct or 0)
    latest = records[0] if records else None
    today_record = next((record for record in records if record.attendance_date == date.today()), None)
    return {
        "student_id": str(profile.user_id),
        "student_name": account.full_name,
        "student_code": profile.student_code,
        "roll_no": profile.roll_no,
        "grade": classroom.grade,
        "section": classroom.section,
        "attendance_percentage": percentage,
        "present_days": present,
        "late_days": late,
        "recorded_days": total,
        "today_status": today_record.status.value if today_record else "not_marked",
        "latest_scan_at": iso(latest.scanned_at) if latest else None,
        "records": [
            {
                "id": str(record.id),
                "date": record.attendance_date.isoformat(),
                "status": record.status.value,
                "source": record.source,
                "scan_uid": record.scan_uid,
                "scanned_at": iso(record.scanned_at),
                "notes": record.notes,
            }
            for record in records[:30]
        ],
    }


@router.post("/scan")
async def record_nfc_scan(payload: NFCScanPayload, db: AsyncSession = Depends(get_db)) -> dict:
    card = await db.scalar(select(RFIDCard).where(RFIDCard.uid == payload.card_uid.strip(), RFIDCard.is_active.is_(True)))
    if not card:
        raise HTTPException(status_code=404, detail="NFC card is not linked to any active student")
    profile = await db.get(StudentProfile, card.student_id)
    account = await db.get(User, card.student_id)
    classroom = await db.get(ClassRoom, profile.class_id) if profile else None
    if not profile or not account or not classroom:
        raise HTTPException(status_code=404, detail="Student profile for this NFC card is missing")

    today = date.today()
    scanned_at = datetime.now(timezone.utc)
    record = await db.scalar(select(AttendanceRecord).where(AttendanceRecord.student_id == card.student_id, AttendanceRecord.attendance_date == today))
    created = record is None
    if record:
        record.status = AttendanceStatus.present
        record.source = "nfc"
        record.scan_uid = card.uid
        record.scanned_at = scanned_at
        record.notes = f"Updated by {payload.device_label}" if payload.device_label else "Updated by NFC scan"
    else:
        record = AttendanceRecord(student_id=card.student_id, class_id=profile.class_id, attendance_date=today, status=AttendanceStatus.present, source="nfc", scan_uid=card.uid, scanned_at=scanned_at, notes=f"Recorded by {payload.device_label}" if payload.device_label else "Recorded by NFC scan")
        db.add(record)

    records_count = await db.scalar(select(AttendanceRecord.id).where(AttendanceRecord.student_id == card.student_id).limit(1))
    if not records_count:
        profile.attendance_pct = Decimal("100.00")
    else:
        all_records = list((await db.scalars(select(AttendanceRecord).where(AttendanceRecord.student_id == card.student_id))).all())
        marked = sum(1 for item in all_records if item.status in (AttendanceStatus.present, AttendanceStatus.late))
        profile.attendance_pct = Decimal(str(round((marked / len(all_records)) * 100, 2))) if all_records else Decimal("100.00")

    await db.commit()
    await db.refresh(record)
    return {"recorded": True, "created": created, "message": "Attendance recorded", "attendance": await serialize_student_attendance(db, profile, account, classroom)}


@router.get("/overview")
async def attendance_overview(student_code: str = "STU-DEMO-001", db: AsyncSession = Depends(get_db)) -> dict:
    rows = await student_rows(db, student_code=student_code)
    if not rows:
        raise HTTPException(status_code=404, detail="Student not found")
    profile, account, classroom = rows[0]
    return await serialize_student_attendance(db, profile, account, classroom)


@router.get("/records")
async def attendance_records(grade: str | None = None, section: str | None = None, db: AsyncSession = Depends(get_db)) -> list[dict]:
    return [await serialize_student_attendance(db, profile, account, classroom) for profile, account, classroom in await student_rows(db, grade=grade, section=section)]


@router.get("/cards")
async def attendance_cards(db: AsyncSession = Depends(get_db)) -> list[dict]:
    rows = await db.execute(select(RFIDCard, StudentProfile, User, ClassRoom).join(StudentProfile, StudentProfile.user_id == RFIDCard.student_id).join(User, User.id == RFIDCard.student_id).join(ClassRoom, ClassRoom.id == StudentProfile.class_id).order_by(User.full_name))
    return [{"uid": card.uid, "label": card.label, "student_name": user.full_name, "student_code": profile.student_code, "grade": classroom.grade, "section": classroom.section, "active": card.is_active} for card, profile, user, classroom in rows.all()]
